sw/host/opentitanlib/src/transport/qemu/transport.rs

```diff
@@ -0,0 +1,105 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::any::Any;
+use std::rc::Rc;
+use std::time::{Duration, Instant};
+
+use anyhow::{ensure, Result};
+use once_cell::sync::{Lazy, OnceCell};
+use regex::Regex;
+use serde_annotate::Annotate;
+
+use crate::io::uart::Uart;
+use crate::transport::qemu::subprocess::{Options, Subprocess};
+use crate::transport::qemu::uart::QemuUart;
+use crate::transport::{
+    Capabilities, Capability, Transport, TransportError, TransportInterfaceType,
+};
+
+/// Represents the qemu transport object.
+pub struct Qemu {
+    subprocess: Subprocess,
+    uart: OnceCell<Rc<QemuUart>>,
+}
+
+impl Qemu {
+    /// Creates a qemu subprocess-hosting transport from `options`.
+    pub fn from_options(options: Options) -> Result<Self> {
+        let uart_open: Lazy<Regex> =
+            Lazy::new(|| Regex::new("uart0.*QEMU waiting for connection").unwrap());
+
+        // Set a deadline for all sockets etc. to be opened.
+        let deadline = Instant::now() + options.timeout;
+
+        // Start the QEMU subprocess.
+        let subprocess = Subprocess::from_options(options)?;
+
+        // Wait for sockets to have opened.
+        subprocess.find(&uart_open, deadline)?;
+
+        Ok(Qemu {
+            subprocess,
+            uart: OnceCell::new(),
+        })
+    }
+
+    /// Shuts down the qemu subprocess.
+    pub fn shutdown(&mut self) -> Result<()> {
+        self.subprocess.kill()
+    }
+}
+
+impl Drop for Qemu {
+    fn drop(&mut self) {
+        self.shutdown().expect("failed to kill qemu subprocess");
+    }
+}
+
+impl Transport for Qemu {
+    fn capabilities(&self) -> Result<Capabilities> {
+        Ok(Capabilities::new(Capability::UART))
+    }
+
+    fn uart(&self, instance: &str) -> Result<Rc<dyn Uart>> {
+        ensure!(
+            instance == "0",
+            TransportError::InvalidInstance(TransportInterfaceType::Uart, instance.to_string())
+        );
+
+        let uart = self
+            .uart
+            .get_or_try_init(|| QemuUart::open(self.subprocess.uart_path()).map(Rc::new))?;
+
+        Ok(uart.clone())
+    }
+
+    fn dispatch(&self, action: &dyn Any) -> Result<Option<Box<dyn Annotate>>> {
+        if let Some(watch) = action.downcast_ref::<Watch>() {
```

Comment:
Is `Watch` used anywhere?

Reply:
I don't understand how `Watch` works, but it's used by `fn dispatch`. I copied this from the verilator module but should have probably tried to understand it first.

Reply:
Deleting `fn dispatch()` and `Watch` doesn't cause any issues, so maybe they're just fine to leave?

The documentation for `dispatch` says:

```rust
/// Invoke non-standard functionality of some Transport implementations.
```

which doesn't help much